src/runtime_src/core/edge/sw_em/generic_pcie_hal2/shim.cxx

```diff
@@ -940,8 +977,18 @@ namespace xclcpuemhal2 {
     if (mLogStream.is_open()) {
       mLogStream << __func__ << ", " << std::this_thread::get_id() << std::endl;
     }
-    if(!sock)
+    if (!sock) {
+      PRINTENDFUNC
+      if (mIsKdsSwEmu && mSWSch && mCore)
+      {
+        mSWSch->fini_scheduler_thread();
+        delete mCore;
+        mCore = nullptr;
+        delete mSWSch;
+        mSWSch = nullptr;
+      }
```

Comment:
Raw new and deletes are not safe.  While I understand these members must be reset, it would be better if the class used std:: smart pointers for these so at least we can be sure they end up getting deleted even if you should miss one such delete.   Smart pointers can be reset using reset() member function.

Reply:
All the necessary checks and deletions handled properly in required places. as you suggested will use smart pointers for this. we will handle this 2020.2.